From: Jeff Moyer <jmoyer@redhat.com>
Date: Fri, 18 Dec 2009 20:36:32 -0500
Subject: [block] cfq: set workload as expired if it doesn't have any slice left
Message-id: <1261168595-15386-9-git-send-email-jmoyer@redhat.com>
Patchwork-id: 22178
O-Subject: [RHEL6 PATCH 08/11] cfq: set workload as expired if it doesn't have
	any slice left
Bugzilla: 548796
RH-Acked-by: Vivek Goyal <vgoyal@redhat.com>

The commit id is from the block tree.  fixes bug 548796.

commit 66ae291978177d5c012015f12b8fbc76dc7d0965
Author: Gui Jianfeng <guijianfeng@cn.fujitsu.com>
Date:   Tue Dec 15 10:08:45 2009 +0100

    cfq: set workload as expired if it doesn't have any slice left

    When a group is resumed, if it doesn't have workload slice left,
    we should set workload_expires as expired. Otherwise, we might
    start from where we left in previous group by error.
    Thanks the idea from Corrado.

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/block/cfq-iosched.c b/block/cfq-iosched.c
index 84290ea..ceb3b43 100644
--- a/block/cfq-iosched.c
+++ b/block/cfq-iosched.c
@@ -2112,7 +2112,9 @@ static void cfq_choose_cfqg(struct cfq_data *cfqd)
 		cfqd->workload_expires = jiffies + cfqg->saved_workload_slice;
 		cfqd->serving_type = cfqg->saved_workload;
 		cfqd->serving_prio = cfqg->saved_serving_prio;
-	}
+	} else
+		cfqd->workload_expires = jiffies - 1;
+
 	choose_service_tree(cfqd, cfqg);
 }
 
